### PR TITLE
[CSL-1201] Guard the state modifying STM action with a mutex

### PR DESCRIPTION
**Pull request by arybczak**
Version of #953 for cardano-sl-0.5-rc.

### CODE REVIEW

File: src/Pos/WorkMode.hs [hunk 1]
```diff
@@ -17,6 +17,8 @@ module Pos.WorkMode
 import           Universum
 
 import           Control.Monad.Fix
+import           Control.Monad.Base             (MonadBase)
+import           Control.Monad.Trans.Control
```
Comment (gromakovsky) [inline src/Pos/WorkMode.hs]: [Our style guide](https://github.com/input-output-hk/cardano-sl-style-guides/blob/master/haskell-style-guide.md#imports) prohibits implicit imports.

File: txp/Pos/Txp/MemState/Class.hs [hunk 2]
```diff
@@ -73,11 +76,15 @@ getMemPool = getTxpLocalData (STM.readTVar . txpMemPool)
 getTxpExtra :: (MonadIO m, MonadTxpMem e m) => m e
 getTxpExtra = getTxpLocalData (STM.readTVar . txpExtra)
 
+txpLocalDataLock :: MVar ()
+txpLocalDataLock = unsafePerformIO $ newMVar ()
```
Comment (gromakovsky) [inline txp/Pos/Txp/MemState/Class.hs]: Is it really necessary to use global variable here? Can't it be added to reader?

File: txp/Pos/Txp/MemState/Class.hs [hunk 3]
```diff
@@ -91,13 +98,15 @@ modifyTxpLocalData f =
         STM.writeTVar txpTip newTip
         STM.writeTVar txpExtra newExtra
         pure res
+ where
+   withLock = L.bracket_ (takeMVar txpLocalDataLock) (putMVar txpLocalDataLock ())
```
Comment (gromakovsky) [inline txp/Pos/Txp/MemState/Class.hs]: We already have two ways of doing `bracket`: `MonadMask` and `Mockable Bracket` (which is bad and we are planning to leave only one). It doesn't make sense to add the third one.

File: src/Pos/WorkMode.hs (hunk 1 above)
Comment (arybczak) [inline src/Pos/WorkMode.hs]: Alright, I didn't know it existed. Do you list and modify them by hand or use some sort of automation (afair GHC can generate explicit import lists, but I don't remember exactly how)? If it's the former, that sounds quite inefficient.

File: txp/Pos/Txp/MemState/Class.hs (hunk 3 above)
Comment (arybczak) [inline txp/Pos/Txp/MemState/Class.hs]: As I mentioned in #953, this function is instantiated with transformer stack that contains `ExceptT`, hence `MonadMask` won't work here. Perhaps `Mockable Bracket` will.

File: src/Pos/WorkMode.hs (hunk 1 above)
Comment (gromakovsky) [inline src/Pos/WorkMode.hs]: Currently we do it by hand, but we have a group of people (actually only two people) working on a separate utility to facilitate work with imports.

> If it's the former, that sounds quite inefficient.

It makes writing code less efficient, but reading code more efficient. The latter is usually more important.

**Pull request merged**